### PR TITLE
Installation guide changes for docker setup.

### PR DESCRIPTION
**Pull request by indyaah**
Fixes jhipster/generator-jhipster#2191

### CODE REVIEW

**Comment by pascalgrimaud** [comment]
1) are you sure that we don't need anymore to add a portmap to the VM ?

2) same question for volumes with svendowideit/samba ?

3) we should add -d (daemon) to docker run (I know, the original didn't have)

4) you specify a container name with --name. We should explain to user how to delete the container, because next time he launch docker run, he will have : `The name "jhipster" is already in use by container e2b70bdddb5f`
Or simply : `docker start jhipster` to use the same container

5) about docker exec, you can use `docker exec -u` or `docker exec --user=` to define a specific user.
You'll be able to go into container with jhipster user instead of root

6) now we can use `docker exec` instead of `ssh`, maybe we should remove openssh-server in Dockerfile

7) don't understand `docker exec -it jhipster /jhipster`. Inside container, you have to `sudo chown jhipster /jhipster`

8) to avoid `sudo chown`, we should use /home/jhipster/app/ for example as a volume inside container, instead of /jhipster
   the command to start is longer but more safe :
   `docker run -d --name=jhipster -v ~/jhipster:/home/jhipster/app -p 8080:8080 -p 3000:3000 -p 3001:3001 -t jdubois/jhipster-docker`

In my opinion, we should improve the actual jhipster-docker :
* remove openssh-server
* create directory app inside container
* change volume from /jhipster to /home/jhipster/app/ (for example)
* maybe add workdir to command run. The user will be in the volume directory, ready to type : `yo jhipster`
    * `docker run -d --name=jhipster --workdir=/home/jhipster/app/ -v ~/jhipster:/home/jhipster/app/ -p 8080:8080 -p 3000:3000 -p 3001:3001 -t jdubois/jhipster-docker` 

@jdubois : are you ok with these 4 changes ?
If yes, @indyaah do you want to work on it and PR ?

Then, update the guide with Toolbox like you did, and it will be good !
(I'll try to find a computer with Windows to test Toolbox!)

**Comment by indyaah** [comment]
@pascalgrimaud 
1) are you sure that we don't need anymore to add a portmap to the VM ?
>> Yes, tested on Mac.

2) same question for volumes with svendowideit/samba ?
>> Yes, again tested mounting volumes directly on mac, works fine.

3) we should add -d (daemon) to docker run (I know, the original didn't have)
>> Yes, I agree.

4) you specify a container name with --name. We should explain to user how to delete the container, because next time he launch docker run, he will have : The name "jhipster" is already in use by container e2b70bdddb5f
Or simply : docker start jhipster to use the same container
>>  I would update.

5) about docker exec, you can use docker exec -u or docker exec --user= to define a specific user.
You'll be able to go into container with jhipster user instead of root
>> I thought user might need to change directory permissions and hence better login as root but I completely agree that if implement item 8, we dont need root login.

6) now we can use docker exec instead of ssh, maybe we should remove openssh-server in Dockerfile
>> Will do.

7) don't understand docker exec -it jhipster /jhipster. Inside container, you have to sudo chown jhipster /jhipster

8) to avoid sudo chown, we should use /home/jhipster/app/ for example as a volume inside container, instead of /jhipster
the command to start is longer but more safe :
docker run -d --name=jhipster -v ~/jhipster:/home/jhipster/app -p 8080:8080 -p 3000:3000 -p 3001:3001 -t jdubois/jhipster-docker
>> Yeah, completely agree. 

I would update the existing PR for documentation and create a new PR on jhipster/jhipster-docker for updated docker image.

Thanks.

**Comment by indyaah** [comment]
@pascalgrimaud : Updated with changes for item 1 to 7.
Working on docker image, once PR for that is approved, will updated item 8 here as well.

**Comment by indyaah** [comment]
Updated for item 8

**Comment by jdubois** [comment]
@pascalgrimaud is this good for you to merge? I didn't have a look at it yet

**Comment by pascalgrimaud** [comment]
I didn't look since the last update. I will review it tomorrow morning

File: installation.html [hunk 1]
```diff
@@ -11,312 +11,216 @@
...
-Start by adding your SSH public key to the Docker container:
+</ul>
 </p>
 <p>
-<code>
-cat ~/.ssh/id_rsa.pub | ssh -p 4022 jhipster@localhost 'mkdir ~/.ssh && cat >> ~/.ssh/authorized_keys'
-</code>
-</p>
+    <code>
+        docker run --name jhipster -v ~/jhipster:/home/jhipster/app:rw -v ~/.m2:/home/jhipster/.m2:rw -p 8080:8080 -p
```
Comment (pascalgrimaud) [inline installation.html]: Add the working dir :
```
docker run --name jhipster -w /home/jhipster/app -v ~/jhipster:/home/jhipster/app:rw -v ~/.m2:/home/jhipster/.m2:rw -p 8080:8080 -p 3000:3000 -p 3001:3001 -d -t jdubois/jhipster-docker
```

The user will be directly in the good directory, when he goes inside container

File: installation.html [hunk 2]
```diff
@@ -11,312 +11,216 @@
...
+    JHipster has a specific <a href="https://github.com/jhipster/jhipster-docker" target="_blank">jhipster-docker
+    project</a>, which provides a <a href="https://www.docker.io/" target="_blank">Docker</a> container.
 </p>
 <p>
-This project makes a Docker "Trusted build" that is available on:
+    This project makes a Docker "Trusted build" that is available on:
 </p>
 <p>
-<a href="https://registry.hub.docker.com/u/jdubois/jhipster-docker/" target="_blank">https://registry.hub.docker.com/u/jdubois/jhipster-docker/</a>
+    <a href="https://registry.hub.docker.com/u/jdubois/jhipster-docker/" target="_blank">https://registry.hub.docker.com/u/jdubois/jhipster-docker/</a>
```
Comment (pascalgrimaud) [inline installation.html]: The official link has changed (with the new hub) :
https://hub.docker.com/r/jdubois/jhipster-docker/

File: installation.html [hunk 3]
```diff
@@ -11,312 +11,216 @@
...
+Also in case you update the docker image (rebuild or pull from dockerhub), it's better to remove existing container, and run the container all over again.
+To do so, first stop the container and then remove it and run again. i.e. :
+<br/>
 <code>
-ssh -p 4022 jhipster@localhost
+    <ol>
+        <li>docker stop jhipster</li>
+        <li>docker rm jhipster</li>
+        <li>docker pull jdubois/jhipster-docker</li>
+        <li>docker run --name jhipster -v ~/jhipster:/home/jhipster/app:rw -v ~/.m2:/home/jhipster/.m2:rw -p 8080:8080
```
Comment (pascalgrimaud) [inline installation.html]: Same, add working dir

File: installation.html [hunk 4]
```diff
@@ -11,312 +11,216 @@
...
 <p>
-Run The docker image, with the following options:
+    Run The docker image, with the following options:
 </p>
 <p>
-  <ul>
-    <li>
-      Forward all ports exposed by docker (8080 for Tomcat, 3000 for BrowserSync from the <code>grunt</code> task, 3001 for the BrowserSync UI, and 22 for SSHD). In the following example we forward the container 22 port to the host 4022 port, to prevent some port conflicts
+<ul>
+    <li>The Docker "/jhipster" folder is shared to the local "~/jhipster" folder</li>
```
Comment (pascalgrimaud) [inline installation.html]: It's `/home/jhipster/app` now, instead of "/jhipster" folder

File: installation.html [hunk 5]
```diff
@@ -11,312 +11,216 @@
...
-You need to fix the jhipster folder to give the "jhipster" user ownership of the directory:
+    You can then go to the /home/jhipster/app directory in your container, and start building your app inside Docker:
 </p>
 <p>
-  <code>
-    ssh -p 4022 jhipster@localhost<br/>
-    sudo chown jhipster /jhipster
-  </code>
+    <code>
+        cd ~/jhipster<br/>
```
Comment (pascalgrimaud) [inline installation.html]: no need, if we add working dir

File: installation.html [hunk 6]
```diff
@@ -11,312 +11,216 @@
...
+<p>
+    <code>
+        docker-machine ip default
+    </code>
+</p>
 </p>
 <p>
-As the generated files are in your shared folder, they will not be deleted if you stop your Docker container. However, if you don't want Docker to keep downloading all the Maven and NPM dependencies every time you start the container, you should commit its state.
+    As the generated files are in your shared folder, they will not be deleted if you stop your Docker container.
+    However, if you don't want Docker to keep downloading all the Maven and NPM dependencies every time you start the
```
Comment (pascalgrimaud) [inline installation.html]: Maven from volume now

File: installation.html [hunk 7]
```diff
@@ -11,312 +11,216 @@
...
-</p>
-<ol>
-  <li>
-    <p>
-      Retrieve the name of the container
-    </p>
-    <p>
-      Here is the resultat of the command `docker ps`. Get the name at the column NAMES.
-      <code>
-      CONTAINER ID        IMAGE                            COMMAND                CREATED             STATUS              PORTS                                                                                            NAMES<br/>
```
Comment (pascalgrimaud) [inline installation.html]: Will be interesting to keep an example of docker ps

File: installation.html [hunk 8]
```diff
@@ -11,312 +11,216 @@
...
-</code>
+    <br/><br/>
+    <b>Please note : </b> On linux, you might need to run the docker command as sudo user if your user is not part of
+    docker group. It's a good idea to add your user to docker group so that you can run docker commands as non-root
+    user. Follow steps on <a href="http://askubuntu.com/a/477554">http://askubuntu.com/a/477554</a> to do so.
 </p>
+
+<h3>Accessing the container</h3>
 <p>
-  If the following exception occurs, you need to change the owner of the /jhipster directory. See next step.
```
Comment (pascalgrimaud) [inline installation.html]: I made some tests. This error can occur when the local directory isn't created !
So we should keep this !

**Comment by pascalgrimaud** [comment]
Some minor changes to be near perfect
@indyaah can you review it ?

**Comment by indyaah** [comment]
@pascalgrimaud : Yes, Sure. I will review.

**Comment by sircelsiussexy** [comment]
Hey guys, it might be worth adding two sections to this page that I think can help:

  * Dockerizing an existing application (nothing too fancy, it's mostly about adding the correct volume to the container)
  * Using different containers to run an app (depending on the `yo` options, I like to run with a `mysql` and an `elasticsearch` container, it should be easy to generalize). I think this adds a lot of interest to using docker.

I can provide both sections if needed.

**Comment by pascalgrimaud** [comment]
@sircelsiussexy  : Dockerizing an application is an interesting thing but you have to take care about :
* allowing debugging with IDE
* allowing to hot reload the code inside container (I didn't test it yet)
* working on link to other container if you use a database in dev profile

At the moment, I use docker for launching database. It's too fast with container !
- mysql, psql, oracle (use non official image), mongodb, cassandra, elasticsearch

Do you have a look to the docker-compose file that is generated since the v2.22.0 ?
http://jhipster.github.io/docker_compose.html

**Comment by sircelsiussexy** [comment]
Yeah, I struggled a bit at first when updating the config files, debugging and live reload, but it turns out to work pretty well (debugging is a pain, things like [this blog post](https://ptmccarthy.github.io/2014/07/24/remote-jmx-with-docker/) helped).

I use also use containers for my database, I've tested a few, including the latest mongo and mysql with the latest generators, works like a charm.

I haven't tested the compose yet, because I've been dockerizing jhipster since before 2.22.0 :sunglasses: However one of the things that I like with my workflow is that I'm able to specify my volumes' locations.

**Pull request merged**